### PR TITLE
At Least Once Delivery on Graceful Shutdown

### PR DESCRIPTION
**Pull request by pulltab**
The following implements at-least-once delivery semantics on graceful shutdown.

We achieve this by:

* Requiring sources / filters to inject Shutdown events into downstream hopper queues when they can ensure that no more events will be placed.

* Restricting source/filter shutdown such that each incident producer has emitted a Shutdown.  This ensures that all possible events have been consumed from each producer.

### CODE REVIEW

File: src/filter/mod.rs [hunk 1]
```diff
@@ -62,15 +63,25 @@ pub trait Filter {
     fn run(
         &mut self,
         recv: hopper::Receiver<metric::Event>,
+        sources: std::vec::Vec<std::string::String>,
```
Comment (blt) [inline src/filter/mod.rs]: I do believe you can make this just `Vec<String>`, no? There are some things auto-used.

File: src/matrix.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    /// Construct a new adjacency matrix.
+    pub fn new() -> Self {
+        Adjacency {
+            edges: Default::default(), 
+        }
+    } 
+
+    /// Adds an outbound edge from a node to another.
+    pub fn add_asymmetric_edge(&mut self, from_str: &str, to_str: &str, metadata: std::option::Option<M>) {
+        let to = std::string::String::from_str(to_str).unwrap();
```
Comment (blt) [inline src/matrix.rs]: I _think_ this can be trimmed down to `String::from_str`.

File: src/filter/mod.rs (hunk 1 above)
Comment (pulltab) [reply src/filter/mod.rs]: Shortened everywhere.

Along with String and Option.

File: src/matrix.rs (hunk 2 above)
Comment (pulltab) [reply src/matrix.rs]: Fixed, and yes it can!

**Comment by blt** [comment]
@pulltab suggest running clippy on this. Latest nightly / clippy build. There's some clones of double references and what not in the new code that it points out.

File: src/bin/cernan.rs [hunk 3]
```diff
@@ -282,34 +270,39 @@ fn main() {
...
     {
         let internal_config = &args.internal;
         config_topology
             .insert(cfg_conf!(internal_config), internal_config.forwards.clone());
+        adjacency_matrix.add_edges(cfg_conf!(internal_config), internal_config.forwards.clone(), None);
     }
     if let Some(ref configs) = args.statsds {
         for (config_path, config) in configs {
-            config_topology.insert(config_path.clone(), config.forwards.clone());
+			config_topology.insert(config_path.clone(), config.forwards.clone());
```
Comment (dparton) [inline src/bin/cernan.rs]: Nitpick: indent introduced

File: src/bin/cernan.rs [hunk 4]
```diff
@@ -282,34 +270,39 @@ fn main() {
...
+            adjacency_matrix.add_edges(config_path.clone(), config.forwards.clone(), None);
         }
     }
     if let Some(ref configs) = args.graphites {
         for (config_path, config) in configs {
-            config_topology.insert(config_path.clone(), config.forwards.clone());
+			config_topology.insert(config_path.clone(), config.forwards.clone());
```
Comment (dparton) [inline src/bin/cernan.rs]: Nitpick: indent introduced

File: src/bin/cernan.rs [hunk 5]
```diff
@@ -282,34 +270,39 @@ fn main() {
...
+            adjacency_matrix.add_edges(config_path.clone(), config.forwards.clone(), None);
         }
     }
     if let Some(ref configs) = args.files {
         for config in configs {
             let config_path = cfg_conf!(config);
-            config_topology.insert(config_path.clone(), config.forwards.clone());
+			config_topology.insert(config_path.clone(), config.forwards.clone());
```
Comment (dparton) [inline src/bin/cernan.rs]: Nitpick: indent introduced

File: src/bin/cernan.rs [hunk 6]
```diff
@@ -581,16 +516,18 @@ fn main() {
...
-            let mut native_server_send = Vec::new();
             let poll = mio::Poll::new().unwrap();
             let (registration, readiness) = mio::Registration::new2();
             populate_forwards(
-                &mut native_server_send,
                 Some(&mut flush_sends),
                 &config.forwards,
-                &cfg_conf!(config),
+                &config_path,
+                //&cfg_conf!(config),
```
Comment (dparton) [inline src/bin/cernan.rs]: Nitpick: comment left behind

Comment (pulltab) [reply src/bin/cernan.rs]: Fixed

File: src/bin/cernan.rs (hunk 5 above)
Comment (pulltab) [reply src/bin/cernan.rs]: Will fix with a rustfmt run.

File: src/bin/cernan.rs (hunk 4 above)
Comment (pulltab) [reply src/bin/cernan.rs]: Will fix with a rustfmt run.

File: src/bin/cernan.rs (hunk 3 above)
Comment (pulltab) [reply src/bin/cernan.rs]: Fixed.

File: src/bin/cernan.rs (hunk 4 above)
Comment (pulltab) [reply src/bin/cernan.rs]: Fixed.

File: src/bin/cernan.rs [hunk 7]
```diff
@@ -763,18 +766,18 @@ fn main() {
...
+    // During shutdown sources will propgate metric::Event::Shutdown
+    // to all of its downstream consumers.
     for (id, source_worker) in sources {
-        println!("Signaling shutdown to {:?}", id);
+        info!("Signaling shutdown to {:?}", id);
         source_worker
             .readiness
             .set_readiness(mio::Ready::readable())
-            .expect("Oops!");
+            .expect("Failed to set readiness!");
```
Comment (blt) [inline src/bin/cernan.rs]: Heh, that's a nicer error message than what I had.

File: src/sink/mod.rs [hunk 8]
```diff
@@ -128,12 +129,17 @@ pub trait Sink {
...
+                            //    all of its events have been processed.
                             //
-                            //    2) The given sink synchronously flushes any
-                            //    internal memory.
-                            return;
+                            // 2) Sources shutdown only after receiving Shutdown
+                            // from each of its
+                            // upstream sources/filters.
+                            total_shutdowns += 1;
+                            if total_shutdowns >= sources.len() {
```
Comment (blt) [inline src/sink/mod.rs]: So cool.

File: src/sink/native.rs [hunk 9]
```diff
@@ -111,10 +112,15 @@ impl Sink for Native {
         // discard point
     }
 
-    fn run(&mut self, recv: hopper::Receiver<metric::Event>) {
+    fn run(
+        &mut self,
+        recv: hopper::Receiver<metric::Event>,
+        sources: std::vec::Vec<String>,
```
Comment (blt) [inline src/sink/native.rs]: This could be shortened to `Vec`.

**Pull request merged**